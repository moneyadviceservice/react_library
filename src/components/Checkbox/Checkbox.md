A check box allowing single values to be selected/deselected.

```jsx
import { Form, Formfield, Checkbox } from '@moneypensionservice/directories';

<Form>
  <Formfield legend="Choose your favorite place:" >
    <Checkbox
      label="Paris"
      value="paris" />
    <Checkbox
      label="Berlin"
      value="berlin" />
    <Checkbox
      label="London"
      value="london" />
  </Formfield>
</Form>
```
