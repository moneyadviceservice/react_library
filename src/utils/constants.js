export const breakpoints = {
  xs: 1,
  sm: 48, // 768px
  md: 64, // 1024px
  lg: 90, // 1440px
  xl: 120, // 1920px
}

export const dimensions = ['xs', 'sm', 'md', 'lg', 'xl']

export const mediaQuery = 'only screen'

export const ALIGN_SELF_MAP = {
  center: 'center',
  end: 'flex-end',
  start: 'flex-start',
  stretch: 'stretch',
}

export const SPACING_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
]

