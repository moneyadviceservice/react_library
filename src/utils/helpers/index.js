export * from './background'
export * from './responsive'
export * from './flexbox'
export * from './styles'
export * from './mediaQueries'
export * from './sizing'
export * from './refs'
