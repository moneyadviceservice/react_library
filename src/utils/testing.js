import { render, act } from '@testing-library/react'
import GlobalTheme from '../components/ThemeProvider'

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalTheme, ...options })

const waitRender = async (ui, options) => {
  const { container, ...rest } = customRender(ui, options)

  await act(async () => {
    await Promise.resolve(container)
  })

  return { container, ...rest }
}

export * from '@testing-library/react'

export { customRender as render, waitRender }
