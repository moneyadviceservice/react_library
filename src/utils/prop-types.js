import PropTypes from 'prop-types'
import { SPACING_SIZES } from './constants'

/** spacing prop.type */
const spacingProp = PropTypes.shape({
  bottom: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  horizontal: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  left: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  right: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  top: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  vertical: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
})

export const genericPropTypes = {
  a11yTitle: PropTypes.string,
  alignSelf: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  margin: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  padding: PropTypes.oneOfType([
    PropTypes.oneOf([...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  border: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
}

export function genericPropsDefaults({
  a11yTitle = null,
  alignSelf = null,
  margin = null,
  padding = null,
  border = null,
} = {}) {
  return {
    a11yTitle: a11yTitle,
    alignSelf: alignSelf,
    margin: margin,
    padding: padding,
    border: border,
  }
}
