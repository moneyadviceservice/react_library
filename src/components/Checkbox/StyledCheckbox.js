import styled from 'styled-components'
import { genericStyles } from '../../utils/helpers'

const Label = styled.label`
  ${genericStyles}
  /** conditional styles */
  ${({ margin }) => !margin && `margin-bottom: 0.75rem;`}
  ${({ disabled, theme }) => disabled && `color: ${theme.colors.disabledText};`}
  ${({ hide }) => !hide && `display: block;`}
  font-size: 1rem;
  line-height: 1.375rem;
  
`

const Field = styled.input`
  margin-right: 5px;
  padding: 0;
`

export { Label, Field }
