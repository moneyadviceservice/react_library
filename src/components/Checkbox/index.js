import React from 'react'
import PropTypes from 'prop-types'
import { genericPropTypes, genericPropsDefaults } from '../../utils/prop-types'
import { Label, Field } from './StyledCheckbox'

const Checkbox = ({
  a11yTitle,
  checked,
  children,
  disabled,
  label,
  name,
  onChange,
  required,
  value,
  ...rest
}) => (
  <Label disabled={disabled} {...rest}>
    <Field
      aria-label={a11yTitle}
      aria-required={required}
      checked={checked}
      disabled={disabled}
      name={name}
      onChange={onChange}
      required={required}
      type="checkbox"
      value={value}
    />
    {children || label}
  </Label>
)

// Documentation
Checkbox.propTypes = {
  /** Whether the component is checked. */
  checked: PropTypes.bool,
  /** Text inside the Label component. Works as the caption for the checkbox element. */
  children: PropTypes.string,
  /** Whether the form control is disabled. */
  disabled: PropTypes.bool,
  /** Caption for the checkbox element. */
  label: PropTypes.string,
  /** Name of the input form control. Submitted with the form as part of a name/value pair. */
  name: PropTypes.string,
  /** On change trigger function event. */
  onChange: PropTypes.func,
  /** Indicates that the user must specify a value for the input. */
  required: PropTypes.bool,
  /** Current value of the form control. Submitted with the form as part of a name/value pair. */
  value: PropTypes.string,
  ...genericPropTypes,
}

Checkbox.defaultProps = {
  disabled: false,
  required: false,
  value: undefined,
  ...genericPropsDefaults(),
}

/** @component */
export { Checkbox }
