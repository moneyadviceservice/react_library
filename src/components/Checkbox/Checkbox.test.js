import React from 'react'
import { render, fireEvent } from '../../utils/testing'
import { Checkbox } from '../Checkbox'

describe('<Checkbox />', () => {
  describe('Snapshots', () => {
    it('should match snapshot without props', () => {
      const { container } = render(<Checkbox />)
      expect(container).toMatchSnapshot()
    })

    it('should match snapshot with disabled prop', () => {
      const { container } = render(<Checkbox disabled />)
      expect(container).toMatchSnapshot()
    })
  })

  describe('Checkbox label', () => {
    const label = 'label text'

    it('should display label prop as Checkbox label', () => {
      const { container } = render(<Checkbox label={label} />)

      expect(container.querySelector('label').textContent).toBe(label)
    })

    it('should display children as Checkbox label', () => {
      const { container } = render(<Checkbox>{label}</Checkbox>)

      expect(container.querySelector('label').textContent).toBe(label)
    })

    it('should apply given color to the label', () => {
      const { container } = render(<Checkbox style={{ color: 'red' }} />)
      expect(container.querySelector('label').style.color).toBe('red')
    })
  })

  describe('Checked status', () => {
    it('should changed the checked status of the input when the label is clicked', () => {
      const { container, getByTestId } = render(
        <Checkbox label="text" data-testid="checkbox" />
      )

      expect(container.querySelector('input').checked).toEqual(false)

      fireEvent.click(getByTestId('checkbox'))

      expect(container.querySelector('input').checked).toEqual(true)
    })
  })

  describe('onChange prop', () => {
    it('should call onChange function when Checkbox is changed', () => {
      const onChangeMock = jest.fn()
      const { getByTestId } = render(
        <Checkbox onChange={onChangeMock} label="text" data-testid="checkbox" />
      )

      fireEvent.click(getByTestId('checkbox'))

      expect(onChangeMock).toHaveBeenCalledTimes(1)
    })
  })

  describe('value prop', () => {
    it('should apply the value property', () => {
      const { container } = render(<Checkbox value="my value" />)

      expect(container.querySelector('input').value).toBe('my value')
    })
  })

  describe('name prop', () => {
    it('should apply the name property', () => {
      const { container } = render(<Checkbox name="my name" />)

      expect(container.querySelector('input').name).toBe('my name')
    })
  })
})
