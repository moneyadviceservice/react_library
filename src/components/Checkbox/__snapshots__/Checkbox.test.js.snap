// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> Snapshots should match snapshot with disabled prop 1`] = `
<div>
  <label
    class="StyledCheckbox__Label-sc-1atc3e8-0 dZCEwt"
    disabled=""
  >
    <input
      aria-required="false"
      class="StyledCheckbox__Field-sc-1atc3e8-1 ggmXuk"
      disabled=""
      type="checkbox"
      value=""
    />
  </label>
</div>
`;

exports[`<Checkbox /> Snapshots should match snapshot without props 1`] = `
<div>
  <label
    class="StyledCheckbox__Label-sc-1atc3e8-0 hGMOrn"
  >
    <input
      aria-required="false"
      class="StyledCheckbox__Field-sc-1atc3e8-1 ggmXuk"
      type="checkbox"
      value=""
    />
  </label>
</div>
`;
